light-clients/ics10-grandpa-cw/src/context.rs

```diff
@@ -73,37 +72,38 @@ impl<'a, H> Context<'a, H> {
 		self.deps.storage
 	}
 
-	pub fn insert_relay_header_hashes(&mut self, headers: &[H256]) {
+	pub fn insert_relay_header_hashes(&mut self, now_ms: u64, headers: &[H256]) {
 		if headers.is_empty() {
 			return
 		}
 
-		let mut xs = GRANDPA_HEADER_HASHES_STORAGE.load(self.storage()).unwrap_or_default();
-		xs.reserve(headers.len());
+		let mut hashes = GRANDPA_HEADER_HASHES_STORAGE.load(self.storage()).unwrap_or_default();
+		hashes.reserve(headers.len());
 		for header in headers {
-			xs.push(*header);
+			hashes.push((now_ms, *header));
 			let _ = GRANDPA_HEADER_HASHES_SET_STORAGE
 				.save(self.storage_mut(), header.0.to_vec(), &())
 				.map_err(|e| {
 					self.log(&format!("error saving hash to set: {:?}", e));
 				});
 		}
-		if xs.len() > GRANDPA_BLOCK_HASHES_CACHE_SIZE {
-			let drained = xs.drain(0..xs.len() - GRANDPA_BLOCK_HASHES_CACHE_SIZE);
+		let expired = now_ms.saturating_sub(HEADER_ITEM_LIFETIME.as_millis() as u64);
+		let key = hashes.binary_search_by_key(&expired, |(t, _)| *t).unwrap_or_else(identity);
+		let len = hashes.len();
+		let remove_up_to_index = key.min(len.saturating_sub(HEADER_ITEM_MIN_COUNT));
+		if remove_up_to_index < len {
```

Comment:
In this case, it means that we need to remove up to `remove_up_to_index`. Let's say 
HEADER_ITEM_MIN_COUNT = 5, and current length is x
if x = 6, then len.saturating_sub(HEADER_ITEM_MIN_COUNT) = 1, so we remove only 1 element, to keep the minimum number of headers
if x = 4, then len.saturating_sub(HEADER_ITEM_MIN_COUNT) = 0, so we don't remove anything, because we haven't reached the minimum number of headers